### PR TITLE
Add `role` config to control postgres ownership of schemas and tables

### LINKED ISSUE #16355: 🚀 Feature: Set ownership for new Postgres schema's
### 🔖 Feature description

Currently backstage creates new postgres schema's by running [`CREATE SCHEMA IF EXISTS new_schema`](https://github.com/backstage/backstage/blob/master/packages/backend-common/src/database/connectors/postgres.ts#L145-L160).

This creates schema's owned by the current user. However, this current user is not always the user which needs to access the schema's (see Context section).

To get around this, it would be great to be able to set the ownership for new schema's. Postgres offers the `AUTHORIZATION` keyword for that: [`CREATE SCHEMA IF EXISTS new_schema AUTHORIZATION role;`](https://www.postgresql.org/docs/current/sql-createschema.html)

This feature request is about adding a config parameter to set the `role` of the newly created schema's.

### 🎤 Context

Our infrastructure is a HashiCorp stack (Nomad, Vault, ...). When we spin up a new allocation (ie. job/container/..) it gets postgres credentials from Vault. The way this works is that Vault provisions postgres with a temporary user and password (`v-username-<random_id>`). These users are member of a parent role.

When migrations are triggered, schema's get created by a random `v-user-..`. This sets the PG Schema ownership to be that specific user:

```
backstage_d=> \dn
                List of schemas
    Name    |              Owner               
------------+----------------------------------
 app        | v-backstage-d-f87FlaQV61pl4MEFoE
 auth       | v-backstage-d-f87FlaQV61pl4MEFoE
 catalog    | v-backstage-d-f87FlaQV61pl4MEFoE
 public     | postgres
 scaffolder | v-backstage-d-f87FlaQV61pl4MEFoE
 search     | v-backstage-d-f87FlaQV61pl4MEFoE
 ```
 
When a new allocation comes up, it gets provisioned with a different user. This user cannot access the schema's owned by the previous user and thus fails. Currently our only way to get around this is to manually reassign the schema's to the parent role.
 
By allowing to set the correct autorizations during schema creation, we could automate this.

### ✌️ Possible Implementation

replacing [this function](https://github.com/backstage/backstage/blob/master/packages/backend-common/src/database/connectors/postgres.ts#L152-L154) with something along the lines of:

```typescript
const ensureSchema = async (database: string) => {
  if(dbConfig.setRole) {
    await admin.raw(`CREATE SCHEMA IF NOT EXISTS ?? AUTHORIZATION ??`, [database, dbConfig.setRole]);
  } else {
    await admin.raw(`CREATE SCHEMA IF NOT EXISTS ??`, [database]);
  }
}
```

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by joeppeeters**
## Hey, I just made a Pull Request!

As described in #16355 this PR adds a configuration option do control the ownership of Schema's and Tables created during migrations.

The config value `role` is available on both the default, as plugin scope.

Given the two users specified in
```sql
CREATE ROLE "backstage"
CREATE ROLE "v-backstage-0" WITH LOGIN PASSWORD "passwd"
GRANT "backstage" TO "v-backstage-0";
```

running backstage with a clean database results in:
```
backstage=# \dn
          List of schemas
      Name      |       Owner
----------------+-------------------
 auth           | v-backstage-0
 catalog        | v-backstage-0
 code-coverage  | v-backstage-0
 entityFeedback | v-backstage-0
 ...
```

With the config value `setRole: 'backstage'` in place, it creates
```
backstage=# \dn
          List of schemas
      Name      |       Owner
----------------+-------------------
 auth           | backstage
 catalog        | backstage
 code-coverage  | backstage
 entityFeedback | backstage
 ...
```
It does so by running `CREATE SCHEMA <schema> AUTHORIZATION <role>` if the config value is set.

The same needs to happen for tables. To ensure that this PR adds a eventListener on the database pool, such that every time a new connection is created, it runs `SET ROLE <role>` as first query.

Without the config this results in:
```
\dt *.*
      Schema       |                     Name                      |    Type     |     Owner
--------------------+-----------------------------------------------+-------------+---------------
 auth               | knex_migrations                               | table       | v-backstage-0
 auth               | knex_migrations_lock                          | table       | v-backstage-0
 auth               | signing_keys                                  | table       | v-backstage-0
 catalog            | final_entities                                | table       | v-backstage-0
 catalog            | knex_migrations                               | table       | v-backstage-0
 catalog            | knex_migrations_lock                          | table       | v-backstage-0
 ...
```
But setting the config changes the owner
```
\dt *.*
                                       List of relations
       Schema       |                     Name                      |    Type     |   Owner
--------------------+-----------------------------------------------+-------------+-----------
 catalog            | final_entities                                | table       | backstage
 catalog            | knex_migrations                               | table       | backstage
 catalog            | knex_migrations_lock                          | table       | backstage
 catalog            | location_update_log                           | table       | backstage
 catalog            | locations                                     | table       | backstage
 ...
```

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

closes #16355

### CODE REVIEW

**Comment by joeppeeters** [comment]
This is my first rodeo with both Backstage as Typescript, hope it makes sense.

Also, I could use some pointers to fix the ` Object literal may only specify known properties, and 'setOwner' does not exist in type 'Config<any>` error.

**Comment by benabineri** [comment]
Stale after a week on PRs is quite short! This feature would be great, we have exactly the same setup and issue with dynamic database users.

File: packages/backend-common/config.d.ts [hunk 1]
```diff
@@ -96,6 +96,8 @@ export interface Config {
        * @default database
        */
       pluginDivisionMode?: 'database' | 'schema';
+      /** Configures the ownership of newly created schemas in pg databases. */
+      setOwner?: string;
```
Comment (benjdlambert) [inline packages/backend-common/config.d.ts:100]: How do you feel about making this just `role` instead?

File: packages/backend-common/config.d.ts [hunk 2]
```diff
@@ -125,6 +127,8 @@ export interface Config {
            * This is merged recursively into the base knexConfig
            */
           knexConfig?: object;
+          /** Configures the ownership of newly created schemas in pg databases. */
+          setOwner?: string;
```
Comment (benjdlambert) [inline packages/backend-common/config.d.ts:131]: Same here?

File: packages/backend-common/src/database/connectors/postgres.ts [hunk 3]
```diff
@@ -35,6 +35,14 @@ export function createPgDatabaseClient(
 ) {
   const knexConfig = buildPgDatabaseConfig(dbConfig, overrides);
   const database = knexFactory(knexConfig);
+
+  const owner = dbConfig.getOptionalString('setOwner');
+
+  if (owner) {
+    database.client.pool.on('createSuccess', (_event: any, pgClient: any) => {
```
Comment (benjdlambert) [inline packages/backend-common/src/database/connectors/postgres.ts:42]: can we remove some of these `any` types`?

File: packages/backend-common/src/database/connectors/postgres.ts [hunk 4]
```diff
@@ -35,6 +35,14 @@ export function createPgDatabaseClient(
...
+      pgClient.query(`SET ROLE ${owner}`, () => {});
```
Comment (benjdlambert) [inline packages/backend-common/src/database/connectors/postgres.ts:43]: Can we do `await client.raw(...)` instead of the empty callback?

**Review by benjdlambert — CHANGES_REQUESTED** [review]

File: packages/backend-common/src/database/connectors/postgres.ts (hunk 4 above)
Comment (joeppeeters) [reply packages/backend-common/src/database/connectors/postgres.ts]: No, the client is not a Knex Object, but a 'raw' connection, which in this case is a [pg.client](https://node-postgres.com/apis/client#clientquery). 

I fixed the typing and added a callback so it doesn't fail silently.

Comment (benjdlambert) [reply packages/backend-common/src/database/connectors/postgres.ts]: https://node-postgres.com/apis/client#promise-api Looks like there's a promise interface though. So you should be able to do something like this:

```ts
  if (role) {
    database.client.pool.on(
      'createSuccess',
      async (_: number, pgClient: Client) => {
        await pgClient.query(`SET ROLE ${role}`);
      },
    );
  }
  ```

File: packages/backend-common/package.json [hunk 5]
```diff
@@ -78,6 +78,7 @@
     "morgan": "^1.10.0",
     "node-fetch": "^2.6.7",
     "node-forge": "^1.3.1",
+    "pg": "^8.9.0",
```
Comment (benjdlambert) [inline packages/backend-common/package.json:81]: Don't think we need this dep either, as it should be fine with just the `@types/pg` dependency.

File: .changeset/mean-toys-itch.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/backend-common': patch
+---
+
+Adds config option to set ownership for newly created schemas and tables in Postgres
```
Comment (benjdlambert) [inline .changeset/mean-toys-itch.md:5]: Probably want to update this to include the new `role` naming and maybe some sample config too? :pray:

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Just two small things left, and a comment about the `promise` interface.

**Comment by benjdlambert** [comment]
Alright - I think this is looking pretty good for now - I just amended the PR to use the same version range as other packages so that we don't get duplicate dependencies too! Thanks! :pray:

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**